test/Validate.CFG.cpp

```diff
@@ -0,0 +1,599 @@
+      "OpSelectionMerge %merge None\n");
+
+  string str = header + nameOps("entry", "bad", make_pair("func", "Main")) +
+               types_consts + "%func    = OpFunction %voidt None %funct\n";
+
+  str += entry >> bad;
+  str += bad >> vector<Block>({def, block1, block2, block3, entry});
+  str += def >> merge;
+  str += block1 >> merge;
+  str += block2 >> merge;
+  str += block3 >> merge;
+  str += merge >> end;
+  str += end;
+  str += "OpFunctionEnd\n";
+
+  CompileSuccessfully(str);
+  ASSERT_EQ(SPV_ERROR_INVALID_CFG, ValidateInstructions());
+  EXPECT_THAT(
+      getDiagnosticString(),
+      MatchesRegex("First block [0-9]\\[entry\\] of funciton [0-9]\\[Main\\] "
+                   "is targeted by block [0-9]\\[bad\\]"));
+}
+
+TEST_F(ValidateCFG, BranchToBlockInOtherFunctionBad) {
+  Block entry("entry");
+  Block middle("middle", SpvOpBranchConditional);
+  Block end("end", SpvOpReturn);
+
+  middle.setBody(
+      "%cond    = OpSLessThan %intt %one %two\n"
+      "OpSelectionMerge %end None\n");
+
+  Block entry2("entry2");
+  Block middle2("middle2");
+  Block end2("end2", SpvOpReturn);
+
+  string str = header + nameOps("middle2", make_pair("func", "Main")) +
+               types_consts + "%func    = OpFunction %voidt None %funct\n";
+
+  str += entry >> middle;
+  str += middle >> vector<Block>({end, middle2});
+  str += end;
+  str += "OpFunctionEnd\n";
+
+  str += "%func2    = OpFunction %voidt None %funct\n";
+  str += entry2 >> middle2;
+  str += middle2 >> end2;
+  str += end2;
+  str += "OpFunctionEnd\n";
+
+  CompileSuccessfully(str);
+  ASSERT_EQ(SPV_ERROR_INVALID_CFG, ValidateInstructions());
+  EXPECT_THAT(
+      getDiagnosticString(),
+      MatchesRegex(
+          "Block\\(s\\) \\{[0-9]\\[middle2\\] .\\} are referenced but not "
+          "defined in function [0-9]\\[Main\\]"));
+}
+
+TEST_F(ValidateCFG, HeaderDoesntDominatesMergeBad) {
+  Block entry("entry");
+  Block merge("merge");
+  Block head("head", SpvOpBranchConditional);
+  Block f("f");
+  Block exit("exit", SpvOpReturn);
+
+  entry.setBody("%cond = OpSLessThan %intt %one %two\n");
+
+  head.setBody("OpSelectionMerge %merge None\n");
+
+  string str = header + nameOps("head", "merge", make_pair("func", "Main")) +
+               types_consts + "%func    = OpFunction %voidt None %funct\n";
+
+  str += entry >> merge;
+  str += head >> vector<Block>({merge, f});
+  str += f >> merge;
+  str += merge >> exit;
+  str += exit;
+
+  CompileSuccessfully(str);
+  ASSERT_EQ(SPV_ERROR_INVALID_CFG, ValidateInstructions());
+  EXPECT_THAT(
+      getDiagnosticString(),
+      MatchesRegex(
+          "Header block [0-9]\\[head\\] doesn't dominate its merge block "
+          "[0-9]\\[merge\\]"));
+}
+
+TEST_F(ValidateCFG, UnreachableMerge) {
+  Block entry("entry");
+  Block branch("branch", SpvOpBranchConditional);
+  Block t("t", SpvOpReturn);
+  Block f("f", SpvOpReturn);
+  Block merge("merge");
+  Block end("end", SpvOpReturn);
```

Comment:
I still think this block is unnecessary (`merge` could return), though the test is still correct.  OK to keep, in the interest of moving this PR along.